Feature polyomino

qucker135 opened the pull request:
### Summary

This PR introduces support for solving exact cover problems involving tiling a rectangular grid with polyomino shapes.

### Motivation

This functionality makes the library more useful in puzzle-solving, game mechanics, and computational geometry contexts.

### New functionalities

 - New `polyomino` module for converting grid + shapes into exact cover constraints and possibilities.
 - `ShapeTransform` enum for controlling allowed transformations of polyomino shapes (rotations, reflections).
 - Unit tests verifying correctness of transformations and solutions.

### Other changes

 - Replaced raw pointer comparison using `==` with `std::ptr::eq()` for safer and more idiomatic pointer equality.
 - Fixed other minor clippy warnings.

### Backward compatibility

Fully backward-compatible — no changes to the public API or functional behavior.

### Testing

 - Added unit tests for proposed module.
 - Existing test suite passes. 
 - [Tarpaulin](https://crates.io/crates/cargo-tarpaulin) reports **99.25%** test coverage for the new module.

### Notes

Feedback is welcome, especially regarding naming conventions and module organization.

qucker135 committed:
Fix clippy warnings

qucker135 committed:
Add polyomino tiling puzzle

declanvk commented:
I'll try to fix the MSRV issue separately and find some time to review this, thanks for putting up the PR!

declanvk commented on src/polyomino.rs:
```diff
@@ -0,0 +1,1142 @@
+//! A [Polyomino tiling puzzle](https://en.wikipedia.org/wiki/Polyomino#Tiling_with_polyominoes)
```
Great docs through this module!

qucker135 replied:
Thanks! :) Do not hesitate, however, to comment on inconsistencies, if you find any.

declanvk commented on src/polyomino.rs:
```diff
@@ -0,0 +1,1142 @@
+
+/// Type representing shape of a single polyomino, encoded as binary mask.
+/// 0 represents an empty cell, 1 represents a filled cell.
+type PShape = Vec<Vec<u8>>;
```
non-blocking: I like the simplicity, but this is going to have a lot of memory overhead compared to a `struct PShape(Vec<u8>)` or `struct PShape(bit_set::BitSet)`. I guess it would really depend on how large these shapes are, specifically shapes with lots of rows will have more `Vec` overhead

Maybe as a followup, we could switch to one of those and  allocate `width * height` elements (either `u8` or individual bits) ahead of time, and then provide convenience function to translate from 2-D points in the shape coordinates back to a index within the container.

Plus, you'd be able to add other impls like a `From<[[u8; W]; H> for PShape` if you make it a new-type versus just a type alias.

qucker135 replied:
You are right; I reimplemented PShape as a new type (struct) instead of type alias. The new type is based on 1D vector type (Vec<u8>) - this representation seemed sufficient for my use case (rather simple puzzles with 8x8, 10x10 grids) and easier to implement for certain functions (e.g. `Polyomino::rotate`) than in `BitSet` case.  I may try to create an implementation based on `BitSet` too, if you find that necessary, though (to what extent reduction of memory consumption is purpose of that library, even if it was connected to manipulation on single bits?).

declanvk commented on src/polyomino.rs:
```diff
@@ -0,0 +1,1142 @@
+                        row.remove(0);
+                    }
+                }
+                shape
```
understanding: What happens when you have a shape like:

```
1 1 0 1 1
1 1 0 1 1
0 0 0 0 0
1 1 0 1 1
1 1 0 1 1
```

Is this a valid polyomino?

qucker135 replied:
Exactly; Polyomino shapes do not necessarily need to be one connected piece - however, asserting, that each border (upper and lowest row, left-most and right-most column) contain at least one `1`, simplifies the latter step of possibilities generation.

declanvk commented on src/polyomino.rs:
```diff
@@ -0,0 +1,1142 @@
+                }
+                // Keep removing empty columns from the left and right
+                // until no more can be removed.
+                for col in (0..shape[0].len()).rev() {
```
clarification: Just to confirm, this top loop removes empty columns from the right, and the `while` loop down below removes empty columns from the left?

declanvk replied:
Related to my non-blocking comment above, I wonder if it would be better to make some of these constraints (like no empty rows/columns on the polyomino borders) as part of the `PShape` construction?

We'd still need to make it a newtype around something. I think its an improvement because we're making the requirements of the individual shapes more concrete, but up for debate.

qucker135 replied:
You read my intention about mentioned loops correctly (the comment probably mislead about the order of the loops - first removed the right column, the second from the left) - however, now this whole section is moved to `PShape::new` and adjusted to the new representation of the type.

qucker135 committed:
Polyomino: Change PShape representation

This commit introduces changes produced in the following discussion:
https://github.com/declanvk/dancing-links/pull/10.
It changes inner representation of PShape (from Vec<Vec<_>> to simple Vec<_>),
treating it as a struct with attached impl block, rather than simple
type alias.
It also moves constraints imposed on that type to `new` function of that
type (type constructor) - earlier assertion of these constraints was
done by `Polyomino` struct, which should not be its responsibility.

declanvk committed:
Bump MSRV to 1.80

New version of `rayon-core` required bumping MSRV

declanvk commented:
Sorry it took me so long to get back to this, I pushed the MSRV bump to this PR, once the tests pass I'll merge! Thanks for making the revisions too

declanvk committed:
Apply clippy --fix

qucker135 commented:
Thanks a lot!
